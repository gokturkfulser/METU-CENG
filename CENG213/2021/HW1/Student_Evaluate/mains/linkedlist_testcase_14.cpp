#include <iostream>

#include "LinkedList.h"
//#include "../LinkedList.h"

template<class T>
void printSizeIncludingDummies(LinkedList<T> *llist) {
    std::cout << "-> Calculating the size including dummies." << std::endl;

    int count = 0;

    Node<T> *temp = llist->getDummyHead();

    while (temp) {
        count += 1;
        temp = temp->next;
    }

    std::cout << "size is " << count << std::endl;
}

template<class T>
void traverseReverse(LinkedList<T> *llist) {
    std::cout << "-> Printing the list in reverse." << std::endl;

    if (llist == NULL) {
        std::cout << "NULL" << std::endl;
        return;
    } else if (llist->isEmpty()) {
        std::cout << "The list is empty." << std::endl;
        return;
    } else {
        Node<T> *node = llist->getActualTail();

        while (node && node->prev) {
            std::cout << *node << std::endl;
            node = node->prev;
        }
    }
}

/*
 * Case 14 : Default constructor; various inserts; various gets; print.
 */
int main() {
    std::cout << "-> Creating a linked list with default constructor." << std::endl;

    LinkedList<int> llist;

    std::cout << "-> Making some insertions to the list." << std::endl;

    llist.insertAtTheBeginning(32);
    llist.insertAtTheEnd(34);
    llist.insertAtTheBeginning(31);
    llist.insertAtTheEnd(35);
    llist.insertBeforeGivenNode(33, llist.getNode(34));

    std::cout << "-> Printing the list." << std::endl;

    llist.traverse();
    traverseReverse(&llist);
    printSizeIncludingDummies(&llist);

    for (int i = 30 ; i <= 36 ; ++i) {
        std::cout << "-> Getting node with data " << i << "." << std::endl;
        Node<int> *temp = llist.getNode(i);
        if (temp == NULL) {
            std::cout << "no node with data " << i << " exists" << std::endl;
        } else {
            std::cout << "exists. data must be " << i << ". it is " << temp->data << std::endl;
        }
    }

    std::cout << "-> Printing the list." << std::endl;

    llist.traverse();
    traverseReverse(&llist);
    printSizeIncludingDummies(&llist);

    return 0;
}
