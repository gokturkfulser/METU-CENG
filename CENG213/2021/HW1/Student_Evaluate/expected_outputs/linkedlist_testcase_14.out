-> Creating a linked list with default constructor.
-> Making some insertions to the list.
-> Printing the list.
31
32
33
34
35
-> Printing the list in reverse.
35
34
33
32
31
-> Calculating the size including dummies.
size is 7
-> Getting node with data 30.
no node with data 30 exists
-> Getting node with data 31.
exists. data must be 31. it is 31
-> Getting node with data 32.
exists. data must be 32. it is 32
-> Getting node with data 33.
exists. data must be 33. it is 33
-> Getting node with data 34.
exists. data must be 34. it is 34
-> Getting node with data 35.
exists. data must be 35. it is 35
-> Getting node with data 36.
no node with data 36 exists
-> Printing the list.
31
32
33
34
35
-> Printing the list in reverse.
35
34
33
32
31
-> Calculating the size including dummies.
size is 7
